### PR TITLE
Add total VMs created metric

### PR DESCRIPTION
**Pull request by machadovilaca**
**What this PR does / why we need it**:

Track the total number of VMs created by virt-controller

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
Add total VMs created metric
```

```jira-ticket
https://issues.redhat.com/browse/CNV-15536
```

/cc @sradco @enp0s3

### CODE REVIEW

File: docs/metrics.md [hunk 1]
```diff
@@ -81,6 +81,9 @@ The current available memory of the VM containers based on the rss. Type: Gauge.
 ### kubevirt_vm_container_free_memory_bytes_based_on_working_set_bytes
 The current available memory of the VM containers based on the working set. Type: Gauge.
 
+### kubevirt_vm_created_total
+Amount of VMs created, broken down by namespace. Type: Counter.
```
Comment (fabiand) [inline docs/metrics.md:85]: Created since when?
Since instal, update, controller pod restart, phase of the moon :) …

Comment (machadovilaca) [reply docs/metrics.md:85]: we can't obviously have the counter track VMs from previous versions, but other than that Prometheus takes care of summing all counter values.

maybe this is a good opportunity to think once again about more metrics metadata, release lifecycle, version when it was added, when they are deprecated, etc... (would also help with the discussion in https://groups.google.com/g/kubevirt-dev/c/7p3q5Lo71hs/m/jIJGNJpdAwAJ)

File: docs/metrics.md [hunk 2]
```diff
@@ -81,6 +81,9 @@ The current available memory of the VM containers based on the rss. Type: Gauge.
...
+
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply docs/metrics.md:86]: providers using KubeVirt (through insights) and users would benefit from this because it would ease the process of understanding the VM usage through time (with absolute values and increase/delta/etc functions to understand trends)

[the remark this answers is not in the document]
Comment (mhenriks) [reply docs/metrics.md:71]: The counter appears to be incremented whenever a VM is:

1.  Created for the first time 
2.  Deleted and recreated
3.  Reconciled the first time by a new instance of virt-controller

On a system with a lot of churn there will be a big drop when virt-controller is restarted.  Maybe that's not a big deal?  How is that handled by consumers?

[the remark this answers is not in the document]
Comment (machadovilaca) [reply docs/metrics.md:71]: Prometheus handles the counter restarting at 0, when `virt-controller` is restarted.

The only issue here would be point 3: "Reconciled the first time by a new instance of virt-controller". But with:

```
// VM is nil or already processed
	if vm == nil || len(vm.Status.Conditions) != 0 {
		return
	}
```

that I added in `func NewVMCreated(vm *v1.VirtualMachine)`, since the VM would have the status conditions set, I would expect it not to happen, am I missing something?

File: pkg/monitoring/virt-controller/metrics/vm_metrics.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+	vmsCreatedCounter = operatormetrics.NewCounterVec(
+		operatormetrics.MetricOpts{
+			Name: "kubevirt_vm_created_total",
+			Help: "Amount of VMs created, broken down by namespace.",
+		},
+		[]string{"namespace"},
+	)
+
+	vmMap = map[string]bool{}
```
Comment (akalenyu) [inline pkg/monitoring/virt-controller/metrics/vm_metrics.go:26]: So we're essentially keeping state in the virt-controller here, which (correct me if I'm wrong) means that the created VM metric will start lying as soon as the virt-controller pod dies?

Comment (machadovilaca) [reply pkg/monitoring/virt-controller/metrics/vm_metrics.go:26]: this state is to prevent "concurrent" requests when the VM is first created
because the `vm.Status.Conditions` are not set yet in these (which would make this function being skipped as I mentioned in https://github.com/kubevirt/kubevirt/pull/10418#discussion_r1364694364)

Comment (akalenyu) [reply pkg/monitoring/virt-controller/metrics/vm_metrics.go:26]: I mean, on a system with existing VMs, when the virt-controller dies because of maintenance or something,
the metric would reset to 0 (lie). Is that not the case?

Comment (machadovilaca) [reply pkg/monitoring/virt-controller/metrics/vm_metrics.go:26]: That is expected with counters, Prometheus handles those counter resets

Comment (akalenyu) [reply pkg/monitoring/virt-controller/metrics/vm_metrics.go:26]: So we would be okay with this metric not reported the correct amount of VMs after a restart?

Comment (machadovilaca) [reply pkg/monitoring/virt-controller/metrics/vm_metrics.go:26]: I wouldn't put it in that way. We want the metric correctly report the number of VMs created by that instance of 'virt-controller'

**Review by enp0s3 — COMMENTED** [review]
@machadovilaca @sradco Hi. My apologies but I find it hard to follow the motivation behind creating this metric. From my POV we can also add metric for VM deletion, VM modification. But what is the use-case we are trying to reach here? why is it important to track the creation rate?

**Comment by machadovilaca** [comment]
> @machadovilaca @sradco Hi. My apologies but I find it hard to follow the motivation behind creating this metric. From my POV we can also add metric for VM deletion, VM modification. But what is the use-case we are trying to reach here? why is it important to track the creation rate?

Currently we have no way of getting an historical information of how many VMs were created in a given cluster.
Yes we have some metrics tracking status change, but those are only good during the Prometheus retention time, after that those number are gone.
This metric would allows us to know the 'real' over-time usage of KubeVirt

**Comment by enp0s3** [comment]
> > @machadovilaca @sradco Hi. My apologies but I find it hard to follow the motivation behind creating this metric. From my POV we can also add metric for VM deletion, VM modification. But what is the use-case we are trying to reach here? why is it important to track the creation rate?
> 
> @enp0s3 Currently we have no way of getting an historical information of how many VMs were created in a given cluster. Yes we have some metrics tracking status change, but those are only good during the Prometheus retention time, after that those number are gone. This metric would allows us to know the 'real' over-time usage of KubeVirt

@machadovilaca Thank you for the reply. I see this as an answer to `what` but not to `why`

**Comment by sradco** [comment]
The `why` is as @enp0s3 mentioned. 
"This metric would allows us to know the 'real' over-time usage of KubeVirt".

**Comment by xpivarc** [comment]
Did we check if any existing Kubernetes metric could be used for this?

**Comment by machadovilaca** [comment]
> Did we check if any existing Kubernetes metric could be used for this?

yes, we could track `virt-launcher` pods or even use KubeVirt VM status metric, the problem with those approaches is that we would lose information previous to retention date

**Comment by machadovilaca** [comment]
@xpivarc

**Review by xpivarc — COMMENTED** [review]
I wonder why don't we implement this in the admitter? The admitter will process each VM exactly once(at least this hold for creation). It seems to me we will get the wrong numbers for the metric in the current implementation.

**Comment by machadovilaca** [comment]
> I wonder why don't we implement this in the admitter? The admitter will process each VM exactly once(at least this hold for creation). It seems to me we will get the wrong numbers for the metric in the current implementation.

@xpivarc good suggesting, I updated the PR to push the metric on vms-admitter

File: pkg/monitoring/virt-controller/metrics/vm_metrics.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+	vmsCreatedCounter = operatormetrics.NewCounterVec(
+		operatormetrics.MetricOpts{
+			Name: "kubevirt_vm_created_total",
+			Help: "Amount of VMs created, broken down by namespace.",
+		},
+		[]string{"namespace"},
+	)
+)
+
+func NewVMCreated(vm *v1.VirtualMachine) {
```
Comment (xpivarc) [inline pkg/monitoring/virt-controller/metrics/vm_metrics.go:22]: Can we reduce the `vm` to `namespace`?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go [hunk 5]
```diff
@@ -210,6 +211,10 @@ func (admitter *VMsAdmitter) Admit(ar *admissionv1.AdmissionReview) *admissionv1
 	reviewResponse := admissionv1.AdmissionResponse{}
 	reviewResponse.Allowed = true
 
+	if ar.Request.Operation == admissionv1.Create {
+		metrics.NewVMCreated(&vm)
```
Comment (xpivarc) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:215]: I think vm might not have the namespace set. Could you check if we populate it?

File: docs/metrics.md [hunk 6]
```diff
@@ -66,6 +66,9 @@ The current available memory of the VM containers based on the rss. Type: Gauge.
 ### kubevirt_vm_container_free_memory_bytes_based_on_working_set_bytes
 The current available memory of the VM containers based on the working set. Type: Gauge.
 
+### kubevirt_vm_created_total
+Amount of VMs created, broken down by namespace. Type: Counter.
```
Comment (xpivarc) [inline docs/metrics.md:70]: As @fabiand pointed out, you might want to elaborate on this.

File: pkg/monitoring/virt-controller/metrics/vm_metrics.go (hunk 4 above)
Comment (machadovilaca) [reply pkg/monitoring/virt-controller/metrics/vm_metrics.go:22]: since we are using a pointer I think we don't have any performance problems with copying the structure, and personally, I think it makes more sense that a function `NewVMCreated` receives a VM as an argument and not a Namespace. Even if in the future we want more labels or create new metrics in the function, it would be easier

File: docs/metrics.md (hunk 6 above)
Comment (machadovilaca) [reply docs/metrics.md:70]: updated

File: pkg/monitoring/virt-api/metrics/vm_metrics.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package metrics
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply pkg/monitoring/virt-api/metrics/vm_metrics.go:1]: added

File: pkg/virt-api/api.go [hunk 8]
```diff
@@ -186,6 +187,12 @@ func (app *virtAPIApp) Execute() {
 	app.ConfigureOpenAPIService()
 	app.reInitChan = make(chan string, 10)
 
+	// setup monitoring
+	err = metrics.SetupMetrics()
+	if err != nil {
+		return
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply pkg/virt-api/api.go:193]: I think you are correct, added a panic

**Comment by fossedihelm** [comment]
@machadovilaca IIUC the metric is reset every time the instance of the virt-api is restarted.
 This means that there is the possibility that the reported value is not correct.
 From this point of view, how can a metric that can report wrong values, be useful?
 I mean, I think that the reported value could not be real, and as consequence, all the related statistics.
 WDYT?
 Thanks!

**Comment by machadovilaca** [comment]
> @machadovilaca IIUC the metric is reset every time the instance of the virt-api is restarted. This means that there is the possibility that the reported value is not correct. From this point of view, how can a metric that can report wrong values, be useful? I mean, I think that the reported value could not be real, and as consequence, all the related statistics. WDYT? Thanks!

@fossedihelm that is the expected behavior and common in Prometheus. Since we are using a counter, which is monotonically increasing, Prometheus has mechanisms to handle when a value goes back to an inferior value (with the restart of the virt-api for example, where we would start counting from zero again), what they call a 'counter restart'

**Comment by fossedihelm** [comment]
/lgtm
Thank you!

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go [hunk 9]
```diff
@@ -210,6 +211,10 @@ func (admitter *VMsAdmitter) Admit(ar *admissionv1.AdmissionReview) *admissionv1
 	reviewResponse := admissionv1.AdmissionResponse{}
 	reviewResponse.Allowed = true
 
+	if ar.Request.Operation == admissionv1.Create {
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: added

**Review by xpivarc — COMMENTED** [review]
/approve

**Review by fossedihelm — APPROVED** [review]
Thanks
